src/lazy/binary/encoded_value.rs

```diff
@@ -0,0 +1,268 @@
+use crate::binary::non_blocking::type_descriptor::Header;
+use crate::types::SymbolId;
+use crate::IonType;
+use std::ops::Range;
+
+/// Type, offset, and length information about the serialized value over which the
+/// NonBlockingRawBinaryReader is currently positioned.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct EncodedValue {
+    // If the compiler decides that a value is too large to be moved/copied with inline code,
+    // it will relocate the value using memcpy instead. This can be quite slow by comparison.
+    //
+    // Be cautious when adding new member fields or modifying the data types of existing member
+    // fields, as this may cause the in-memory size of `EncodedValue` instances to grow.
+    //
+    // See the Rust Performance Book section on measuring type sizes[1] for more information.
+    // [1] https://nnethercote.github.io/perf-book/type-sizes.html#measuring-type-sizes
+
+    // The type descriptor byte that identified this value; includes the type code, length code,
+    // and IonType.
+    pub header: Header,
+
+    // Each encoded value has up to five components, appearing in the following order:
+    //
+    // [ field_id? | annotations? | header (type descriptor) | header_length? | value ]
+    //
+    // Components shown with a `?` are optional.
+    //
+    // EncodedValue stores the offset of the type descriptor byte from the beginning of the
+    // data source (`header_offset`). The lengths of the other fields can be used to calculate
+    // their positions relative to the type descriptor byte. For example, to find the offset of the
+    // field ID (if present), we can do:
+    //     header_offset - annotations_header_length - field_id_length
+    //
+    // This allows us to store a single `usize` for the header offset, while other lengths can be
+    // packed into a `u8`. Values are not permitted to have a field ID or annotations that take
+    // more than 255 bytes to represent.
+    //
+    // We store the offset for the header byte because it is guaranteed to be present for all values.
+    // Field IDs and annotations appear earlier in the stream but are optional.
+
+    // The number of bytes used to encode the field ID (if present) preceding the Ion value. If
+    // `field_id` is undefined, `field_id_length` will be zero.
+    pub field_id_length: u8,
+    // If this value is inside a struct, `field_id` will contain the SymbolId that represents
+    // its field name.
+    pub field_id: Option<SymbolId>,
+    // The number of bytes used to encode the annotations wrapper (if present) preceding the Ion
+    // value. If `annotations` is empty, `annotations_header_length` will be zero.
+    pub annotations_header_length: u8,
+    // The number of bytes used to encode the series of symbol IDs inside the annotations wrapper.
+    pub annotations_sequence_length: u8,
+    // Type descriptor byte location.
+    pub header_offset: usize,
```

Comment:
This is the only field that is _not_ relative; the `header_offset` is its absolute offset in the input stream. I'll modify the comment to clarify that.

Reply:
This is also addressed in dev comments above:

```rust
    // EncodedValue stores the offset of the type descriptor byte from the beginning of the
    // data source (`header_offset`). The lengths of the other fields can be used to calculate
    // their positions relative to the type descriptor byte. For example, to find the offset of the
    // field ID (if present), we can do:
    //     header_offset - annotations_header_length - field_id_length
    //
    // This allows us to store a single `usize` for the header offset, while other lengths can be
    // packed into a `u8`. Values are not permitted to have a field ID or annotations that take
    // more than 255 bytes to represent.
    //
    // We store the offset for the header byte because it is guaranteed to be present for all values.
    // Field IDs and annotations appear earlier in the stream but are optional.
```